kaggle_environments/envs/lux_ai_2022/luxai2022/unit.py

```diff
@@ -0,0 +1,121 @@
+from dataclasses import dataclass
+from enum import Enum
+from typing import List
+
+import numpy as np
+try:
+    from termcolor import colored
+except:
+    pass
+from luxai2022.config import EnvConfig, UnitConfig
+
+from luxai2022.globals import TERM_COLORS
+from luxai2022.map.position import Position
+from luxai2022.team import FactionTypes, Team
+
+class UnitType(Enum):
+    LIGHT = "Light"
+    HEAVY = "Heavy"
+    
+@dataclass
+class UnitCargo:
+    ice: int = 0
+    ore: int = 0
+    water: int = 0
+    metal: int = 0
+    def state_dict(self):
+        return dict(
+            ice=self.ice,
+            ore=self.ore,
+            water=self.water,
+            metal=self.metal
+        )
+
+class Unit:
+    def __init__(self, team: Team, unit_type: UnitType, unit_id: str, env_cfg: EnvConfig) -> None:
+        self.unit_type = unit_type
+        self.team_id = team.team_id
+        self.team = team
+        self.unit_id = unit_id
+        self.pos = Position(np.zeros(2, dtype=int))
+        
+        self.cargo = UnitCargo()
+        # TODO - replace with a deque perhaps?
+        self.action_queue: List = []
+        self.unit_cfg: UnitConfig = env_cfg.ROBOTS[unit_type.name]
+        self.power = env_cfg.ROBOTS[unit_type.name].INIT_POWER
+        self.cargo_space = env_cfg.ROBOTS[unit_type.name].CARGO_SPACE
+        self.battery_capacity = env_cfg.ROBOTS[unit_type.name].BATTERY_CAPACITY
+
+    def __str__(self) -> str:
+        out = f"[{self.team_id}] {self.unit_id} {self.unit_type} at {self.pos}"
+        if TERM_COLORS:
+            return colored(out, self.team.faction.value.color)
+        return out
+    def is_heavy(self) -> bool:
+        return self.unit_type == UnitType.HEAVY
+    def next_action(self):
+        if len(self.action_queue) == 0: return None
+        action = self.action_queue.pop(0)
+        if action.repeat:
+            self.action_queue.append(action)
+        return action
+    def move_power_cost(self, rubble_at_target: int):
+        return self.unit_cfg.MOVE_COST + self.unit_cfg.RUBBLE_MOVEMENT_COST * rubble_at_target
+    def state_dict(self):
+        return dict(
+            team_id=self.team_id,
+            unit_id=self.unit_id,
+            power=self.power,
+            unit_type=self.unit_type.name,
+            pos=self.pos.pos,
+            cargo=self.cargo.state_dict(),
+            action_queue=[a.state_dict() for a in self.action_queue],
+        )
+    def add_resource(self, resource_id, amount):
+        if amount < 0: amount = 0
+        if resource_id == 0:
+            transfer_amount = min(self.cargo_space - self.cargo.ice, amount)
```

Comment:
As written, doesn't this allow `cargo_space` of each resource to be carried? didn't we want to enforce this limit across everything you are carrying? or are you somehow enforcing that there should only be one resource type carried? (which I think we want the ability to carry multiple types)

Reply:
Yes, I think we were allowing them to carry `cargo_space` amount of each of the 4 resources. Thought this was a little bit easier to handle so players don't need to worry about transferring out resources as often in order to free up room. 

Might also offer some more intricate transferring by allowing an agent to hold multiple resources and transfer all these types out (sort of like a 1-tile sized one-way cross-section)